docs/proposals/manifest-hydrator.md

```diff
@@ -0,0 +1,364 @@
+
+Example output might look like this:
+
+````markdown
+# dev-west Manifests
+
+[manifest.yaml](./manifest.yaml) contains the hydrated manifests for the dev-west application.
+
+These are the details of the most recent change;
+* Author: Michael Crenshaw <michael@example.com>
+* Message: chore: bumped image tag to v0.0.2
+* Time: 2024-03-27 10:32:04 UTC
+
+To reproduce the manifest hydration, do the following:
+
+```
+git clone https://github.com/argoproj/argocd-example-apps
+cd argocd-example-apps
+git checkout ab2382f
+kustomize edit set image my-app:v0.0.2
+kustomize build environments/dev/west
+```
+````
+
+The hydrator will also write a `hydrator.metadata` file containing a JSON representation of all the values available for README templating. This metadata can be used by external systems (e.g. a PR-based promoter system) to generate contextual information about the hydrated manifest's provenance.
+
+```json
+{
+  "commands": ["kustomize edit set image my-app:v0.0.2", "kustomize build ."],
+  "drySHA": "ab2382f",
+  "commitAuthor": "Michael Crenshaw <michael@example.com>",
+  "commitMessage": "chore: bump Helm dependency chart to 32.1.12",
+  "repoURL": "https://github.com/argoproj/argocd-example-apps"
+}
+```
+
+To request a commit to the hydrated branch, the application controller will make a gRPC call to the CommitManifests service.
+
+A single call will bundle all the changes destined for a given targetBranch.
+
+It's the application controller's job to ensure that the user has write access to the repo before making the call.
+
+```protobuf
+// CommitManifests represents the caller's request for some Kubernetes manifests to be pushed to a git repository.
+message CommitManifests {
+  // repoURL is the URL of the repo we're pushing to. HTTPS or SSH URLs are acceptable.
+  required string repoURL = 1;
+  // targetBranch is the name of the branch we're pushing to.
+  required string targetBranch = 2;
+  // drySHA is the full SHA256 hash of the "dry commit" from which the manifests were hydrated.
+  required string drySHA = 3;
+  // commitAuthor is the name of the author of the dry commit.
+  required string commitAuthor = 4;
+  // commitMessage is the short commit message from the dry commit.
+  required string commitMessage = 5;
+  // commitTime is the dry commit timestamp.
+  required string commitTime = 6;
+  // details holds the information about the actual hydrated manifests.
+  repeated CommitPathDetails details = 7;
+}
+
+// CommitManifestDetails represents the details about a 
+message CommitPathDetails {
+  // path is the path to the directory to which these manifests should be written.
+  required string path = 1;
+  // manifests is a list of JSON documents representing the Kubernetes manifests.
+  repeated string manifests = 2;
+  // readme is a string which will be written to a README.md alongside the manifest.yaml. 
+  required string readme = 3;
+}
+
+message CommitManifestsResponse {
+}
+```
+
+### Use cases
```

Comment:
- Describe behavior and configuration for different common usage of ArgoCD
    - [ ] ApplicationSet defined in Git
    - [ ] Applications generated in Git
    - [ ] Applications created/managed dynamically
    - [ ] Monorepo of applications vs single repo
    - [ ] multiple path in same hydrated branch.